RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc

```diff
@@ -0,0 +1,637 @@
+    nnlsWork_.pulseCovArray[iBX]   = FullSampleMatrix::Constant(0);
+
+    if (offset==100) {
+      nnlsWork_.ampVec.coeffRef(iBX) = sqrt(nnlsWork_.pedConstraint.coeff(0));
+    }
+    else {
+
+      updatePulseShape(nnlsWork_.amplitudes.coeff(nnlsWork_.tsOffset + offset), 
+		       nnlsWork_.pulseShapeArray[iBX], 
+		       nnlsWork_.pulseDerivArray[iBX],
+		       nnlsWork_.pulseCovArray[iBX]);
+      
+      nnlsWork_.ampVec.coeffRef(iBX)=0;
+
+      nnlsWork_.pulseMat.col(iBX) = nnlsWork_.pulseShapeArray[iBX].segment(nnlsWork_.fullTSOffset - offset, nnlsWork_.tsSize);
+
+    }
+  }
+
+  nnlsWork_.pulseMat.col(nnlsWork_.nPulseTot-1) = SampleVector::Ones(nnlsWork_.tsSize);
+
+  nnlsWork_.aTaMat.resize(nnlsWork_.nPulseTot, nnlsWork_.nPulseTot);
+  nnlsWork_.aTbVec.resize(nnlsWork_.nPulseTot);
+
+  double chiSq = minimize(); 
+
+  nnlsWork_.residuals = nnlsWork_.pulseMat*nnlsWork_.ampVec - nnlsWork_.amplitudes;
+
+  bool foundintime = false;
+  unsigned int ipulseintime = 0;
+
+  for (unsigned int iBX=0; iBX<nnlsWork_.nPulseTot; ++iBX) {
+    if (nnlsWork_.bxs.coeff(iBX)==0) {
+      ipulseintime = iBX;
+      foundintime = true;
+    }
+  }
+
+  correctedOutput.clear();
+  if (foundintime) {
+    correctedOutput.push_back(nnlsWork_.ampVec.coeff(ipulseintime)); //charge
+    double arrivalTime = calculateArrivalTime();
+    correctedOutput.push_back(arrivalTime); //time
+    correctedOutput.push_back(chiSq); //chi2
+
+  }
+  
+}
+
+double MahiFit::minimize() const {
+
+  int iter = 0;
+  double oldChiSq=999;
+  double chiSq=oldChiSq;
+
+  while (true) {
+    if (iter>=nMaxItersMin_) {
+      break;
+    }
+
+    updateCov();
+
+    if (nnlsWork_.nPulseTot>1) {
+      nnls();
+    }
+    else {
+      onePulseMinimize();
+    }
+
+    double newChiSq=calculateChiSq();
+    double deltaChiSq = newChiSq - chiSq;
+
+    if (newChiSq==oldChiSq && newChiSq<chiSq) {
+      break;
+    }
+    oldChiSq=chiSq;
+    chiSq = newChiSq;
+
+    if (std::abs(deltaChiSq)<deltaChiSqThresh_) break;
+
+    iter++;
+    
+  }
+
+  return chiSq;
+
+}
+
+void MahiFit::updatePulseShape(double itQ, FullSampleVector &pulseShape, FullSampleVector &pulseDeriv,
+			       FullSampleMatrix &pulseCov) const {
+  
+  float t0=meanTime_;
+  if (applyTimeSlew_) 
+    t0+=HcalTimeSlew::delay(std::max(1.0, itQ), slewFlavor_);
+
+  nnlsWork_.pulseN.fill(0);
+  nnlsWork_.pulseM.fill(0);
+  nnlsWork_.pulseP.fill(0);
+
+  const double xx[4]={t0, 1.0, 0.0, 3};
+  const double xxm[4]={-nnlsWork_.dt+t0, 1.0, 0.0, 3};
+  const double xxp[4]={ nnlsWork_.dt+t0, 1.0, 0.0, 3};
+
+  (*pfunctor_)(&xx[0]);
+  psfPtr_->getPulseShape(nnlsWork_.pulseN);
+
+  (*pfunctor_)(&xxm[0]);
+  psfPtr_->getPulseShape(nnlsWork_.pulseM);
+  
+  (*pfunctor_)(&xxp[0]);
+  psfPtr_->getPulseShape(nnlsWork_.pulseP);
+  for (unsigned int iTS=nnlsWork_.fullTSOffset; iTS<nnlsWork_.fullTSOffset + nnlsWork_.tsSize; iTS++) {
+    if (nnlsWork_.tsSize==8) {
```

Comment:
Will this still work for other values of tsSize except for 8 and 10 ?
If not, there better be an assert.
This if-else and the one below could be simpler if a correction is defined ```delta = tsSize == 8 ? 1 : 0```.